10_rest/willshen8/pkg/puppy/rest.go

```diff
@@ -0,0 +1,124 @@
+package puppy
+
+import (
+	"net/http"
+	"strconv"
+
+	"github.com/go-chi/chi"
+	"github.com/go-chi/render"
+)
+
+// RestHandler provides a puppy storer to store the puppies
+type RestHandler struct {
+	store Storer
+}
+
+// NewRestHandler is a factory method that makes new storer
+func NewRestHandler(storer Storer) *RestHandler {
+	return &RestHandler{store: storer}
+}
+
+// HandleGet gets the puppy by id and displays the results.
+func (rh *RestHandler) HandleGet(w http.ResponseWriter, r *http.Request) {
+	id, err := strconv.Atoi(chi.URLParam(r, "id"))
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		http.Error(w, http.StatusText(400)+": "+ErrorInvalidInput.String(), 400)
+		return
+	}
+	puppy, readErr := rh.store.ReadPuppy(uint32(id))
+	if readErr != nil {
+		w.WriteHeader(http.StatusNotFound)
+		http.Error(w, http.StatusText(404)+": "+ErrorNotFound.String(), 404)
+		return
+	}
+	render.JSON(w, r, puppy)
+}
+
+// HandlePost post a new puppy and then display the results.
+func (rh *RestHandler) HandlePost(w http.ResponseWriter, r *http.Request) {
+	var p Puppy
+	if decodeErr := render.DecodeJSON(r.Body, &p); decodeErr != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		http.Error(w, http.StatusText(400)+": "+ErrorInvalidInput.String(), 400)
+		return
+	}
+	_, createErr := rh.store.CreatePuppy(&p)
+	if createErr != nil {
+		http.Error(w, http.StatusText(400)+": "+ErrorInvalidInput.String(), 400)
+		return
+	}
+	render.Status(r, http.StatusCreated)
+	render.JSON(w, r, p)
+}
+
+// HandlePut updates the existing puppy with new fields in the payload.
+func (rh *RestHandler) HandlePut(w http.ResponseWriter, r *http.Request) {
+	id, err := strconv.Atoi(chi.URLParam(r, "id"))
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		http.Error(w, http.StatusText(400)+": "+ErrorInvalidInput.String(), 400)
+		return
+	}
+	var p Puppy
+	if decodeErr := render.DecodeJSON(r.Body, &p); decodeErr != nil {
+		w.WriteHeader(http.StatusUnprocessableEntity)
+		http.Error(w, http.StatusText(http.StatusUnprocessableEntity)+": "+ErrorInvalidInput.String(),
+			http.StatusUnprocessableEntity)
+		return
+	}
+
+	updateErr := rh.store.UpdatePuppy(uint32(id), &p)
+	if updateErr != nil {
+		// if updateError is due to invalid id
+		if updateErr.Error() == "Error code 1: Puppy ID can't be found, update operation failed." {
```

Comment:
string comparison is not ideal here - can we compare by error code?

Reply:
I have made generalised error codes based on classifications, but error messages are more specific to give users better feedbacks. So I am comparing both the code and string. Happy to change it if there is a more suitable way of doing it.